core/src/al/miniaudio_context.cpp

```diff
@@ -0,0 +1,369 @@
+#define MINIAUDIO_IMPLEMENTATION
+#include <cubos/core/al/miniaudio_context.hpp>
+#include <cubos/core/log.hpp>
+#include <cubos/core/reflection/external/string.hpp>
+
+using namespace cubos::core::al;
+
+class MiniaudioBuffer : public impl::Buffer
+{
+public:
+    ma_decoder decoder;
+
+    MiniaudioBuffer(const void* data, size_t dataSize)
+    {
+        if (ma_decoder_init_memory(data, dataSize, nullptr, &decoder) != MA_SUCCESS)
+        {
+            CUBOS_ERROR("Failed to initialize Decoder from data");
+        }
+
+        mValid = true;
+    }
+
+    ~MiniaudioBuffer() override
+    {
+        ma_decoder_uninit(&decoder);
+    }
+
+    size_t length() override
+    {
+        ma_uint64 lengthInPCMFrames;
+        ma_result result = ma_decoder_get_length_in_pcm_frames(&decoder, &lengthInPCMFrames);
+
+        if (result != MA_SUCCESS)
+        {
+            CUBOS_ERROR("Failed to get the length of audio in PCM frames.");
+            return 0;
+        }
+
+        // Calculate the length in seconds: Length in PCM frames divided by the sample rate.
+        return static_cast<size_t>(lengthInPCMFrames) / decoder.outputSampleRate;
+    }
+
+    bool isValid() const
+    {
+        return mValid;
+    }
+
+private:
+    bool mValid = false;
+};
+
+class MiniaudioSource : public impl::Source
+{
+public:
+    MiniaudioSource(ma_engine& engine)
+        : mEngine(engine)
+    {
+    }
+
+    ~MiniaudioSource() override
+    {
+        ma_sound_uninit(&mSound);
+    }
+
+    void setBuffer(Buffer buffer) override
+    {
+        // Try to dynamically cast the Buffer to a MiniaudioBuffer.
+        auto miniaudioBuffer = std::dynamic_pointer_cast<MiniaudioBuffer>(buffer);
+
+        if (miniaudioBuffer == nullptr)
+        {
+            CUBOS_FAIL("Buffer is not of type MiniaudioBuffer.");
+            return;
+        }
```

Comment:
```suggestion
        // Try to dynamically cast the Buffer to a MiniaudioBuffer.
        auto miniaudioBuffer = std::static_pointer_cast<MiniaudioBuffer>(buffer);
```
I think you can safely assume there won't be other buffer - we probably don't need to cover a scenario where there are more than one audio context implementations coexisting in the same application.